### PR TITLE
NUT-17: WebSocket updates

### PR DESCRIPTION
**Pull request by Egge21M**
# Fixes: none

## Description

This adds WebSocket management, as well as a couple of methods on CashuWallet for full NUT-17 support

## Changes

- Added WSConnection Class
- Handle automatic connects and reconnects
- Added Message Queue
- Added CashuWallet methods
- Added unit and integration tests

## PR Tasks

- [x] Open PR
- [x] run `npm run test` --> no failing unit tests
- [x] run `npm run format`

### CODE REVIEW

File: src/CashuWallet.ts [hunk 1]
```diff
@@ -863,6 +867,153 @@ class CashuWallet {
...
+				if (p.state === MintQuoteState.PAID) {
+					callback(p);
+				}
+			},
+			errorCallback
+		);
+		return () => {
+			this.mint.webSocketConnection?.cancelSubscription(subId, callback);
+		};
+	}
```
Comment (callebtc) [inline src/CashuWallet.ts:954]: Can this not use `onMintQuoteUpdates`? Seems like a lot of repeated boiler plate.

Comment (Egge21M) [reply src/CashuWallet.ts]: Very good point! Done

File: src/WSConnection.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	JsonRpcReqParams,
+	RpcSubId
+} from './model/types';
+import { OnOpenError, OnOpenSuccess } from './model/types/wallet/websocket';
+import { getWebSocketImpl } from './ws';
+
+export class WSConnection {
+	public readonly url: URL;
+	private readonly _WS: typeof WebSocket;
+	private ws: WebSocket | undefined;
```
Comment (alexlwn123) [inline src/WSConnection.ts:15]: Should these private fields be prefixed by `_`? like the classes in utils?

File: src/WSConnection.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	private connectionPromise: Promise<void> | undefined;
+	private subListeners: { [subId: string]: Array<(payload: any) => any> } = {};
+	private rpcListeners: { [rpsSubId: string]: any } = {};
```
Comment (alexlwn123) [inline src/WSConnection.ts:18]: What's the purpose of having both subListeners and rpcListeners?

File: src/WSConnection.ts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+						return;
+					}
+					if (this.subListeners[subId]?.length > 0) {
+						const notification = parsed as JsonRpcNotification;
+						this.subListeners[subId].forEach((cb) => cb(notification.params.payload));
+					}
+				}
+			}
+		} catch (e) {
+			console.log(e);
```
Comment (alexlwn123) [inline src/WSConnection.ts:141]: Should this try to call the errorCallback?

File: src/model/types/index.ts [hunk 5]
```diff
@@ -17,3 +17,48 @@ export type InvoiceData = {
...
+type JsonRpcParams = any;
+
+export type JsonRpcReqParams = {
+	kind: RpcSubKinds;
+	filters: Array<string>;
+	subId: string;
+};
+
+type JsonRpcSuccess<T = any> = {
+	jsonrpc: '2.0';
```
Comment (alexlwn123) [inline src/model/types/index.ts:34]: maybe make jsonrpc `readonly`, to make sure the editor doesn't display the type as `string`

**Review by gandlafbtc — COMMENTED** [review]
There is a lot to this PR, I might need some more time to go through it in more detail, but I also don't wanna bike-shed it. On the surface it LGTM. So I'd say after fixing these minor issues, we put it in an RC.

File: src/WSConnection.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+	closeSubscription(subId: string) {
+		this.ws?.send(JSON.stringify(['CLOSE', subId]));
+	}
+
+	addSubListener(subId: string, callback: (payload: any) => any) {
+		(this.subListeners[subId] = this.subListeners[subId] || []).push(callback);
+	}
+
+	//TODO: Move to RPCManagerClass
```
[the remark this answers is not in the document]
Comment (Egge21M) [reply src/WSConnection.ts:99]: Yeah, i wanted to modularise this further to help with testing. It's not necessary for functionality, but only there for clean-code purposes. IMO it's fine to leave this here for someone to pick up at some point

File: src/WSConnection.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+						return;
+					}
+					if (this.subListeners[subId]?.length > 0) {
+						const notification = parsed as JsonRpcNotification;
+						this.subListeners[subId].forEach((cb) => cb(notification.params.payload));
+					}
+				}
+			}
+		} catch (e) {
+			console.log(e);
```
[the remark this answers is not in the document]
Comment (Egge21M) [reply src/WSConnection.ts:162]: Good point, will update. I don't think we should actually log anything to the console, unless the API consumer opts into it. I will open an issue to implement a "verbose" flag or similar

**Pull request merged**